textfrag/src/types/directive/textdirective.rs

```diff
@@ -0,0 +1,448 @@
+    /// - `TextFragmentError::PercentDecodeError`, if the percent decode fails for the directive
+    ///
+    pub fn from_fragment_as_str(fragment: &str) -> Result<TextDirective, TextFragmentError> {
+        // If text directive delimiter (text=) is not found, return error
+        if !fragment.contains(TEXT_DIRECTIVE_DELIMITER) {
+            return Err(TextFragmentError::NotTextDirective);
+        }
+
+        // XXX: we anticipate no issues in constructing the regex - otherwise, let there be panic!
+        let regex = Regex::new(TEXT_DIRECTIVE_REGEX);
+        if regex.is_err() {
+            log::error!(
+                "Error constructing the regex object: {}",
+                TEXT_DIRECTIVE_REGEX
+            );
+            return Err(TextFragmentError::RegexConsructionError(
+                TEXT_DIRECTIVE_REGEX.to_string(),
+            ));
+        }
+
+        let regex = regex.unwrap();
+        if let Ok(Some(result)) = regex.captures(fragment) {
+            let start = result
+                .name("start")
+                .map(|start| start.as_str())
+                .unwrap_or_default();
+            let start = TextDirective::percent_decode(start)?;
+
+            // Start is MANDATORY - check for valid directive input
+            if start.is_empty() {
+                return Err(TextFragmentError::StartDirectiveMissingError);
+            }
+
+            let mut search_kind = TextDirectiveKind::Start;
+
+            let prefix = result
+                .name("prefix")
+                .map(|m| m.as_str())
+                .unwrap_or_default();
+            let prefix = TextDirective::percent_decode(prefix)?;
+            if !prefix.is_empty() {
+                search_kind = TextDirectiveKind::Prefix;
+            }
+
+            let end = result.name("end").map(|e| e.as_str()).unwrap_or_default();
+            let end = TextDirective::percent_decode(end)?;
+
+            let suffix = result
+                .name("suffix")
+                .map(|m| m.as_str())
+                .unwrap_or_default();
+            let suffix = TextDirective::percent_decode(suffix)?;
+
+            Ok(TextDirective {
+                prefix,
+                start,
+                end,
+                suffix,
+                raw_directive: fragment.to_owned(),
+                status: RefCell::new(TextDirectiveStatus::NotStarted),
+                search_kind: search_kind.into(),
+                next_offset: Cell::new(0),
+                resultant_str: RefCell::new(String::new()),
+            })
+        } else {
+            Err(TextFragmentError::RegexCaptureError(
+                fragment.to_string(),
+                TEXT_DIRECTIVE_REGEX.to_string(),
+            ))
+        }
+    }
+
+    /// [Internal] the use of regular expression does not comply with the specification
+    /// To be used for testing purposes only
+    fn _check(&self, input: &str) -> Result<FragmentDirectiveStatus, TextFragmentError> {
```

Comment:
Using underscores for internal methods is rather uncommon. It's clear enought that there's no `pub` in front.

```suggestion
    fn check(&self, input: &str) -> Result<FragmentDirectiveStatus, TextFragmentError> {
```

Reply:
changed the function name and also to suppress the warning, i've included the #[allow(dead_code)] directive

Reply:
renamed the function per feedback